docs/bsp.md

```diff
@@ -578,27 +579,51 @@ object BuildTargetEventKind {
 The `BuildTargetEventKind` information can be used by clients to trigger reindexing or update the
 user interface with the new information.
 
-### Build Target Text Documents Request
+### Sources Request
 
-The build target text documents request is sent from the client to the server to query for the list of source files that are part of a given list of build targets.
+The build target sources request is sent from the client to the server to query for the list of text documents and directories that are belong to a build target.
+The sources response must not include sources that are external to the workspace, see `buildTarget/dependencySources`.
 
-* method: `buildTarget/textDocuments`
-* params: `BuildTargetTextDocumentsParams`
+* method: `buildTarget/sources`
+* params: `BuildTargetSourcesParams`
 
 ```scala
-trait BuildTargetTextDocumentsParams {
+trait SourcesParams {
   def targets: List[BuildTargetIdentifier]
 }
 ```
 
 Response:
 
-* result: `BuildTargetTextDocumentsResult`, defined as follows
+* result: `BuildTargetSourcesResult`, defined as follows
 
 ```scala
-trait BuildTargetTextDocumentsResult {
-  /** The source files used by this target */
-  def textDocuments: List[TextDocumentIdentifier]
+trait SourcesResult {
+  def items: List[SourcesItem]
+}
+trait SourcesItem {
+  def target: BuildTargetIdentifer
+  /** The text documents or and directories that belong to this build target. */
+  def sources: List[SourceItem]
+}
+trait SourceItem {
+  /** Either a text document or a directory. A directory entry must end with a forward
+    * slash "/" and a directory entry implies that every nested text document within the 
+    * directory belongs to this source item.
+    */
+  def uri: Uri
+
+  /** The category of this source item, must be one of SourceItemKind. */
+  def kind: Int
+}
+object SourceItemKind {
+  /** Source code that is manually edited by the user. */
+  val Source = 1
+  /** Data files that are manually managed by the user. */
+  val Resource = 2
```

Comment:
:thinking: Why do we need this kind if we already have `buildTarget/resources` to get the resources of a build target?
  
I suggest we remove `Resource` and use rename `Source` to `Default` and `Excluded` to `Generated` (@jastice seems to propose the same thing in a comment below). What do you think?

Reply:
Good point, see https://github.com/scalacenter/bsp/pull/47#discussion_r228447419